localstack-core/localstack/services/lambda_/provider.py

```diff
@@ -971,6 +972,51 @@ def create_function(
             version, return_qualified_arn=False, return_update_status=False
         )
 
+    def _validate_runtime(self, package_type, runtime):
+        runtimes = ALL_RUNTIMES
+        if config.LAMBDA_RUNTIME_VALIDATION:
+            runtimes = list(itertools.chain(RUNTIMES_AGGREGATED.values()))
+
+        if package_type == PackageType.Zip and runtime not in runtimes:
+            # deprecated runtimes have different error
+            if runtime in DEPRECATED_RUNTIMES:
+                HINT_LOG.info(
+                    "Set env variable LAMBDA_RUNTIME_VALIDATION to 0"
+                    " in order to allow usage of deprecated runtimes"
+                )
+                self._check_for_recomended_migration_target(runtime)
+
+            raise InvalidParameterValueException(
+                f"Value {runtime} at 'runtime' failed to satisfy constraint: Member must satisfy enum value set: {VALID_RUNTIMES} or be a valid ARN",
+                Type="User",
+            )
+
+    def _check_for_recomended_migration_target(self, deprecated_runtime):
+        # AWS offers recommended runtime for migration for "newly" deprecated runtimes
+        # in order to preserve parity with error messages we need the code bellow
+
+        match deprecated_runtime:
+            case Runtime.go1_x:
+                latest_runtime = RUNTIMES_AGGREGATED["provided"][0]
+            case Runtime.nodejs12_x:
+                latest_runtime = Runtime.nodejs18_x
+            case Runtime.dotnetcore3_1:
+                latest_runtime = RUNTIMES_AGGREGATED["dotnet"][0]
+            case _:
+                match = re.match(r"^[a-zA-Z]+", deprecated_runtime)
+                runtime_type = match.group(0)
+
+                if runtime_type in RUNTIMES_AGGREGATED:
+                    latest_runtime = RUNTIMES_AGGREGATED[runtime_type][0]
+                else:
+                    latest_runtime = None
```

Comment:
Not a big fan of making this dependent on the order of items in `RUNTIMES_AGGREGATED`, since it is not always latest to newest ordered, and we make an assumption that the suggestions might change with the release of new runtimes.

Maybe introduce a mapping of `deprecated-runtime -> suggestion`, similar to our `RUNTIMES_AGGREGATED`, in runtimes.py? This way we could have a simple lookup in a dict, and have every suggestion on one place. We should regenerate the snapshots of the suggestion exception after a new runtime as well, to see if it changes. Might be worth adding those tests to the runtime marker, WDYT @joe4dev ?

Reply:
I love these suggestions.
I'll add them :)

Thank you.

Reply:
I've added a marker and `DEPRECATED_RUNTIMES_UPGRADES` dictionary